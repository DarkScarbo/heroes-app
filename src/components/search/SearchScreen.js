import React, { useMemo } from 'react';
import queryString from 'query-string';
import { HeroCard } from '../heroes/HeroCard';
import { useForm } from '../../hooks/useForm';
import { useLocation } from 'react-router-dom';
import { getHeroesByName } from '../../selectors/getHeroesByName';

export const SearchScreen = ({ history }) => {

    const location = useLocation();
    const { q = '' } = queryString.parse( location.search );
    const [ formValues, handleInputChange ] = useForm({ searchText: q });
    const { searchText } = formValues;
    const heroesFiltered = useMemo(() => getHeroesByName( q ), [q])

    const handleSearch = (e) => {
        e.preventDefault();
        history.push(`?q=${ searchText }`);
    }

    return (
        <div>

            <div className='card text-center ms-3 animate__animated animate__fadeIn text-white bg-dark mb-3'>

                <div className='card-body'>

                    <form className="container mt-3" onSubmit={ handleSearch }>

                         <div className="input-group mb-3">

                            <input 
                                type="text"
                                placeholder="Hero name"
                                className="form-control"
                                name="searchText"
                                autoComplete="off"
                                value={ searchText }
                                onChange={ handleInputChange }
                            />

                            <div className="input-group-append">

                                <button 
                                    className="btn btn-primary" 
                                    type="submit">
                                        Search!
                                </button>

                            </div>

                        </div>

                    </form>     
                    
                </div>

            </div>

            { 
                (q !=='' && heroesFiltered.length === 0 ) 
                    && 
                    <div className="alert alert-danger">
                        There are no heroes with { q }
                    </div>
            }

            <div className='card-columns'>
                
                {
                    heroesFiltered.map( hero => (
                        <HeroCard 
                            key={ hero.id }
                            { ...hero }
                        />
                    ))
                }

            </div>
                
        </div>
    )
}
