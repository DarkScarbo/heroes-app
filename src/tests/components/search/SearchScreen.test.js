import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter, Route } from 'react-router-dom';
import { SearchScreen } from '../../../components/search/SearchScreen';


describe('Tests in <SearchScreen />', () => {

    test('should render the component correctly with default values', () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <Route path="/search" component={ SearchScreen } />
            </MemoryRouter>
        );

        expect( wrapper ).toMatchSnapshot();
    })

    test('should show a hero by giving a queryString', () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=Green%20Lantern']}>
                <Route path="/search" component={ SearchScreen } />
            </MemoryRouter>
        );

        expect( wrapper.find('input').prop('value') ).toBe('Green Lantern');
        expect( wrapper ).toMatchSnapshot();

    })

    test('should show an error message if it does not find a hero', () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman123']}>
                <Route path="/search" component={ SearchScreen } />
            </MemoryRouter>
        );

        expect( wrapper.find('.alert-danger').text().trim() ).toBe('There are no heroes with batman123');
        expect( wrapper ).toMatchSnapshot();

    })
    
    test('should call history.push()', () => {

        const history = {
            push: jest.fn()
        };

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman123']}>
                <Route 
                    path="/search" 
                    component={ () => <SearchScreen history={ history } /> } 
                />
            </MemoryRouter>
        );

        wrapper.find('input').simulate('change', {
            target: {
                name: 'searchText',
                value: 'batman'
            }
        });

        wrapper.find('form').prop('onSubmit')({
            preventDefault(){}
            });

        expect( history.push ).toHaveBeenCalledWith(`?q=batman`);

    })
    
})
