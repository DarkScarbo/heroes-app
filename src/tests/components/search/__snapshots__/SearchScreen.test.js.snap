// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <SearchScreen /> should render the component correctly with default values 1`] = `
<div>
  <div
    className="card text-center ms-3 animate__animated animate__fadeIn text-white bg-dark mb-3"
  >
    <div
      className="card-body"
    >
      <hr />
      <h4>
         Find your heroes 
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Hero name"
          type="text"
          value=""
        />
        <button
          className="btn mt-2 btn-block btn-outline-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
  </div>
  <div
    className="card-columns"
  />
</div>
`;

exports[`Tests in <SearchScreen /> should show a hero by giving a queryString 1`] = `
<div>
  <div
    className="card text-center ms-3 animate__animated animate__fadeIn text-white bg-dark mb-3"
  >
    <div
      className="card-body"
    >
      <hr />
      <h4>
         Find your heroes 
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Hero name"
          type="text"
          value="Green Lantern"
        />
        <button
          className="btn mt-2 btn-block btn-outline-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
  </div>
  <div
    className="card-columns"
  >
    <div
      className="card ms-3 animate__animated animate__fadeIn"
      style={
        Object {
          "maxHeight": 540,
          "maxWidth": 540,
        }
      }
    >
      <div
        className="row no-gutters"
      >
        <div
          className="col-md-4"
        >
          <img
            alt="Green Lantern"
            className="card-img"
            src="./assets/heroes/dc-green.jpg"
          />
        </div>
        <div
          className="col-md-8"
        >
          <div
            className="card-body"
          >
            <h5
              className="card-title"
            >
               
              Green Lantern
               
            </h5>
            <p
              className="card-text"
            >
               
              Alan Scott
               
            </p>
            <p
              className="card-text"
            >
              <small
                className="text-muted"
              >
                 
                All-American Comics #16
                 
              </small>
            </p>
            <a
              href="/hero/dc-green"
              onClick={[Function]}
            >
              More...
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Tests in <SearchScreen /> should show an error message if it does not find a hero 1`] = `
<div>
  <div
    className="card text-center ms-3 animate__animated animate__fadeIn text-white bg-dark mb-3"
  >
    <div
      className="card-body"
    >
      <hr />
      <h4>
         Find your heroes 
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Hero name"
          type="text"
          value="batman123"
        />
        <button
          className="btn mt-2 btn-block btn-outline-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
  </div>
  <div
    className="card-columns"
  >
    <div
      className="alert alert-danger"
    >
      There are no heroes with 
      batman123
    </div>
  </div>
</div>
`;
